Meta/Lagom/Tools/CodeGenerators/GMLCompiler/main.cpp

```diff
@@ -0,0 +1,304 @@
+        generator.appendln(text);
+        return {};
+    };
+
+    generator.append(TRY(String::repeated(' ', (indentation - 1) * 4)).bytes_as_string_view());
+    generator.appendln("{");
+    if (use_object_constructor == UseObjectConstructor::Yes)
+        TRY(append(generator, "@object_name@ = TRY(@class_name@::try_create());"));
+    else
+        TRY(append(generator, "@object_name@ = TRY(adopt_nonnull_ref_or_enomem(new (nothrow) @class_name@()));"));
+
+    // Properties
+    TRY(gml_object.try_for_each_property([&](StringView key, NonnullRefPtr<GUI::GML::JsonValueNode> value) -> ErrorOr<void> {
+        auto value_code = TRY(generate_initializer_for(key, value));
+        if (is_ui_dimension_property(key)) {
+            if (auto ui_dimension = GUI::UIDimension::construct_from_json_value(value); ui_dimension.has_value())
+                value_code = TRY(ui_dimension->as_cpp_source());
+            else
+                // FIXME: propagate precise error cause
+                return Error::from_string_view("UI dimension invalid"sv);
+        } else {
+            // Wrap value in an extra constructor call if necessary.
+            if (auto type = map_property_to_type(key); type.has_value())
+                value_code = TRY(String::formatted("{} {{ {} }}", type.release_value(), value_code));
+        }
+
+        auto property_generator = generator.fork();
+        property_generator.set("key", key);
+        property_generator.set("value", value_code.bytes_as_string_view());
+        TRY(append(property_generator, R"~~~(@object_name@->set_@key@(@value@);)~~~"));
+        return {};
+    }));
+    generator.appendln("");
+
+    // Layout
+    if (gml_object.layout_object() != nullptr) {
+        TRY(append(generator, "RefPtr<GUI::Layout> layout;"));
+        TRY(generate_loader_for_object(*gml_object.layout_object(), generator.fork(), TRY(String::from_utf8("layout"sv)), indentation + 1, UseObjectConstructor::Yes));
+        TRY(append(generator, "@object_name@->set_layout(layout.release_nonnull());"));
+        generator.appendln("");
+    }
+
+    // Children
+    size_t current_child_index = 0;
+    auto next_child_name = [&]() {
+        return String::formatted("{}_child_{}", object_name, current_child_index++);
+    };
+    TRY(gml_object.try_for_each_child_object([&](auto const& child) -> ErrorOr<void> {
+        auto child_generator = generator.fork();
+        auto child_variable_name = TRY(next_child_name());
+        child_generator.set("child_variable_name", child_variable_name.bytes_as_string_view());
+        child_generator.set("child_class_name", child.name());
+        TRY(append(child_generator, "RefPtr<@child_class_name@> @child_variable_name@;"));
+        TRY(generate_loader_for_object(child, child_generator.fork(), child_variable_name, indentation + 1, UseObjectConstructor::Yes));
+
+        // Handle the current two special cases of child adding.
```

Comment:
Should this have a FIXME?

Reply:
If you want to make add_child virtual, sure! I don't think so.